src/people.rs

```diff
@@ -0,0 +1,339 @@
+pub struct PersonCreatedEvent {
+    pub person_id: PersonId,
+}
+
+// Emitted when a person property is updated
+// These should not be emitted outside this module
+#[derive(Copy, Clone)]
+#[allow(clippy::manual_non_exhaustive)]
+pub struct PersonPropertyChangeEvent<T: PersonProperty> {
+    pub person_id: PersonId,
+    pub current: T::Value,
+    pub previous: T::Value,
+}
+
+pub type PersonBuilderCallback = dyn FnOnce(&mut Context, PersonId);
+pub struct PersonBuilder<'a> {
+    /// Internal
+    callbacks: Vec<Box<PersonBuilderCallback>>,
+    /// Reference to the simulation context
+    context: &'a mut Context,
+    /// Id of the person being built
+    person_id: PersonId,
+}
+
+impl<'a> PersonBuilder<'a> {
+    pub fn new(context: &'a mut Context) -> PersonBuilder<'a> {
+        let people_data_container = context.get_data_container_mut(PeoplePlugin);
+        let person_id = PersonId {
+            id: people_data_container.population,
+        };
+        PersonBuilder {
+            context,
+            person_id,
+            callbacks: Vec::new(),
+        }
+    }
+
+    /// Returns a reference to `Context` which might be needed to build the person
+    pub fn get_context(&mut self) -> &mut Context {
+        self.context
+    }
+
+    /// Returns the identifier struct for the person
+    #[must_use]
+    pub fn get_person_id(&self) -> PersonId {
+        self.person_id
+    }
+
+    fn add_callback(&mut self, callback: impl FnOnce(&mut Context, PersonId) + 'static) {
+        self.callbacks.push(Box::new(callback));
+    }
+
+    /// Sets the value a person property of type `T`
+    ///
+    /// Returns `self` to allow for chaining
+    #[must_use]
+    pub fn set_person_property<T: PersonProperty + 'static>(
+        mut self,
+        property: T,
+        value: T::Value,
+    ) -> PersonBuilder<'a> {
+        self.add_callback(move |context, person_id| {
+            context.set_person_property(person_id, property, value);
+        });
+        self
+    }
+
+    /// Inserts the finalized person into the data container after person properties
+    /// have been set.
+    #[must_use]
+    pub fn insert(self) -> PersonId {
+        let people_data_container = self.context.get_data_container_mut(PeoplePlugin);
+        people_data_container.population += 1;
+        let person_id = self.get_person_id();
+        for callback in self.callbacks {
+            callback(self.context, person_id);
+        }
+        self.context.emit_event(PersonCreatedEvent {
+            person_id: self.person_id,
+        });
+        self.person_id
+    }
+}
+
+pub trait ContextPeopleExt {
+    /// Creates a person by instantiating a `PersonBuilder`, which can be used to
+    /// set properties before inserting it into the data container.
+    fn create_person(&mut self) -> PersonBuilder;
+
+    /// Given a Persionid, returns the value of a defined person property
+    fn get_person_property<T: PersonProperty + 'static>(
```

Comment:
Should we make this panic if the `PersonId` hasn't been created yet? (And same for the set method below)